File: backend/routes/predict.js
const router = require('express').Router();
const path = require('path');
const fs = require('fs');
const { spawn } = require('child_process');
const auth = require('../middleware/authMiddleware');
const Prediction = require('../models/Prediction');

const pythonBinary = process.env.PYTHON_BIN || (process.platform === 'win32' ? 'python' : 'python3');

// Disease guidance map matching your Python model's 38 classes
const guidance = {
  'Apple___Apple_scab': {
    description: 'Fungal disease causing dark, scabby lesions on leaves and fruit.',
    suggestions: ['Remove infected leaves', 'Apply fungicide', 'Improve air circulation']
  },
  'Apple___Black_rot': {
    description: 'Fungal disease causing leaf spots and fruit rot.',
    suggestions: ['Prune infected branches', 'Apply appropriate fungicide', 'Remove mummified fruit']
  },
  'Apple___Cedar_apple_rust': {
    description: 'Fungal disease causing orange spots on leaves.',
    suggestions: ['Remove nearby cedar trees if possible', 'Apply fungicide in spring', 'Use resistant varieties']
  },
  'Apple___healthy': {
    description: 'Leaf appears healthy with no signs of disease.',
    suggestions: ['Maintain proper watering', 'Continue regular monitoring']
  },
  'Blueberry___healthy': {
    description: 'Blueberry leaf is healthy.',
    suggestions: ['Continue proper care']
  },
  'Cherry_(including_sour)___Powdery_mildew': {
    description: 'Fungal disease causing white powdery spots on leaves.',
    suggestions: ['Remove infected leaves', 'Improve air circulation', 'Apply sulfur-based fungicide']
  },
  'Cherry_(including_sour)___healthy': {
    description: 'Leaf appears healthy.',
    suggestions: ['Continue good care practices']
  },
  'Corn_(maize)___Cercospora_leaf_spot Gray_leaf_spot': {
    description: 'Fungal disease causing gray leaf spots.',
    suggestions: ['Rotate crops', 'Use resistant varieties', 'Apply fungicide if severe']
  },
  'Corn_(maize)___Common_rust_': {
    description: 'Fungal disease with reddish-brown pustules.',
    suggestions: ['Plant resistant hybrids', 'Apply fungicide if necessary']
  },
  'Corn_(maize)___Northern_Leaf_Blight': {
    description: 'Fungal disease causing long gray-green lesions.',
    suggestions: ['Use resistant varieties', 'Rotate crops', 'Remove crop debris']
  },
  'Corn_(maize)___healthy': {
    description: 'Corn leaf is healthy.',
    suggestions: ['Maintain proper nutrition and watering']
  },
  'Grape___Black_rot': {
    description: 'Fungal disease causing black spots and fruit rot.',
    suggestions: ['Remove infected fruit', 'Apply fungicide', 'Prune for air circulation']
  },
  'Grape___Esca_(Black_Measles)': {
    description: 'Fungal disease causing leaf discoloration and wood decay.',
    suggestions: ['Remove infected wood', 'No cure available - manage symptoms', 'Avoid stress to vines']
  },
  'Grape___Leaf_blight_(Isariopsis_Leaf_Spot)': {
    description: 'Fungal disease with dark leaf spots.',
    suggestions: ['Apply copper fungicide', 'Remove infected leaves', 'Improve drainage']
  },
  'Grape___healthy': {
    description: 'Grape leaf is healthy.',
    suggestions: ['Continue regular care']
  },
  'Orange___Haunglongbing_(Citrus_greening)': {
    description: 'Bacterial disease causing yellow shoots and misshapen fruit.',
    suggestions: ['Remove infected trees', 'Control psyllid insects', 'Use certified disease-free plants']
  },
  'Peach___Bacterial_spot': {
    description: 'Bacterial disease causing dark spots on leaves and fruit.',
    suggestions: ['Use copper sprays', 'Plant resistant varieties', 'Avoid overhead irrigation']
  },
  'Peach___healthy': {
    description: 'Peach leaf is healthy.',
    suggestions: ['Maintain good practices']
  },
  'Pepper,_bell___Bacterial_spot': {
    description: 'Bacterial disease with dark leaf spots.',
    suggestions: ['Use disease-free seeds', 'Apply copper bactericide', 'Rotate crops']
  },
  'Pepper,_bell___healthy': {
    description: 'Pepper leaf is healthy.',
    suggestions: ['Continue proper care']
  },
  'Potato___Early_blight': {
    description: 'Fungal disease with concentric ring patterns on leaves.',
    suggestions: ['Apply fungicide', 'Remove infected leaves', 'Rotate crops']
  },
  'Potato___Late_blight': {
    description: 'Serious fungal disease causing rapid leaf death.',
    suggestions: ['Apply fungicide immediately', 'Remove infected plants', 'Improve air circulation']
  },
  'Potato___healthy': {
    description: 'Potato leaf is healthy.',
    suggestions: ['Monitor regularly', 'Maintain proper care']
  },
  'Raspberry___healthy': {
    description: 'Raspberry leaf is healthy.',
    suggestions: ['Continue regular care']
  },
  'Soybean___healthy': {
    description: 'Soybean leaf is healthy.',
    suggestions: ['Maintain good practices']
  },
  'Squash___Powdery_mildew': {
    description: 'Fungal disease with white powdery coating.',
    suggestions: ['Apply fungicide', 'Improve air circulation', 'Remove infected leaves']
  },
  'Strawberry___Leaf_scorch': {
    description: 'Fungal disease causing leaf edges to brown.',
    suggestions: ['Remove infected leaves', 'Apply fungicide', 'Improve drainage']
  },
  'Strawberry___healthy': {
    description: 'Strawberry leaf is healthy.',
    suggestions: ['Maintain proper care']
  },
  'Tomato___Bacterial_spot': {
    description: 'Bacterial disease causing dark spots.',
    suggestions: ['Use resistant varieties', 'Apply copper spray', 'Avoid overhead watering']
  },
  'Tomato___Early_blight': {
    description: 'Fungal disease with target-like spots.',
    suggestions: ['Remove affected leaves', 'Apply fungicide', 'Mulch to prevent soil splash']
  },
  'Tomato___Late_blight': {
    description: 'Devastating fungal disease.',
    suggestions: ['Apply fungicide', 'Remove infected plants', 'Ensure good air flow']
  },
  'Tomato___Leaf_Mold': {
    description: 'Fungal disease causing yellowing and mold.',
    suggestions: ['Reduce humidity', 'Improve ventilation', 'Apply fungicide']
  },
  'Tomato___Septoria_leaf_spot': {
    description: 'Fungal disease with circular spots.',
    suggestions: ['Remove infected leaves', 'Apply fungicide', 'Avoid wetting foliage']
  },
  'Tomato___Spider_mites Two-spotted_spider_mite': {
    description: 'Pest damage causing stippling and webbing.',
    suggestions: ['Spray with water', 'Use insecticidal soap', 'Introduce predatory mites']
  },
  'Tomato___Target_Spot': {
    description: 'Fungal disease with concentric ring spots.',
    suggestions: ['Apply fungicide', 'Remove infected leaves', 'Improve air circulation']
  },
  'Tomato___Tomato_Yellow_Leaf_Curl_Virus': {
    description: 'Viral disease causing leaf curling and yellowing.',
    suggestions: ['Control whitefly vectors', 'Remove infected plants', 'Use resistant varieties']
  },
  'Tomato___Tomato_mosaic_virus': {
    description: 'Viral disease causing mottled leaves.',
    suggestions: ['Remove infected plants', 'Sanitize tools', 'Use resistant varieties']
  },
  'Tomato___healthy': {
    description: 'Tomato leaf is healthy.',
    suggestions: ['Continue good practices', 'Monitor regularly']
  }
};

// POST /api/predict
router.post('/', auth, async (req, res) => {
  try {
    const { imageUrl } = req.body || {};
    
    console.log('📥 Received prediction request:', { imageUrl, userId: req.user?.id });
    
    if (!imageUrl || !imageUrl.startsWith('/uploads/')) {
      return res.status(400).json({ message: 'Invalid or missing imageUrl' });
    }

    const absPath = path.join(__dirname, '..', imageUrl);
    
    // Check if file exists
    if (!fs.existsSync(absPath)) {
      console.error('❌ File not found:', absPath);
      return res.status(404).json({ message: 'Image file not found' });
    }

    console.log('🐍 Running Python prediction on:', absPath);

    // Run Python script
    const pyPath = path.join(__dirname, '..', 'model', 'predict.py');
    const py = spawn(pythonBinary, [pyPath, '--image', absPath]);

    let stdout = '';
    let stderr = '';

    py.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    py.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    py.on('close', async (code) => {
      if (code !== 0) {
        console.error('❌ Python error (exit code ' + code + '):', stderr);
        return res.status(500).json({ 
          message: 'Prediction failed', 
          error: stderr || 'Python script exited with error' 
        });
      }

      let result;
      try {
        console.log('📤 Python output:', stdout);
        result = JSON.parse(stdout);
      } catch (e) {
        console.error('❌ Invalid JSON from Python:', stdout);
        return res.status(500).json({ message: 'Invalid response from model' });
      }

      const diseaseName = result.disease || 'Unknown';
      const confidence = typeof result.confidence === 'number' ? result.confidence : 0;

      const meta = guidance[diseaseName] || {
        description: 'No description available.',
        suggestions: ['Consult agronomy resources for treatment.']
      };

      // Save prediction to database
      try {
        const pred = await Prediction.create({
          userId: req.user.id,
          imageUrl,
          diseaseName,
          confidence
        });

        console.log('✅ Prediction saved:', pred._id);

        return res.json({
          id: pred._id,
          imageUrl,
          disease: diseaseName,
          confidence,
          description: meta.description,
          suggestions: meta.suggestions
        });
      } catch (dbError) {
        console.error('⚠️ Database save failed:', dbError.message);
        // Still return prediction even if DB save fails
        return res.json({
          imageUrl,
          disease: diseaseName,
          confidence,
          description: meta.description,
          suggestions: meta.suggestions,
          warning: 'Prediction not saved to database'
        });
      }
    });
  } catch (err) {
    console.error('❌ Predict route error:', err);
    return res.status(500).json({ 
      message: 'Server error', 
      error: err.message 
    });
  }
});

module.exports = router;

File: backend/model/predict.py
#!/usr/bin/env python3
import argparse
import json
import os
import sys

# Print Python version for debugging
print(f"DEBUG: Python version: {sys.version}", file=sys.stderr)
print(f"DEBUG: Script started", file=sys.stderr)

# Suppress TensorFlow warnings
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

try:
    import numpy as np
    print(f"DEBUG: numpy imported successfully", file=sys.stderr)
except Exception as e:
    print(json.dumps({"error": f"Failed to import numpy: {str(e)}"}))
    sys.exit(2)

try:
    from tensorflow import keras
    print(f"DEBUG: tensorflow.keras imported successfully", file=sys.stderr)
except Exception as e:
    print(json.dumps({"error": f"Failed to import tensorflow: {str(e)}"}))
    sys.exit(2)

try:
    from PIL import Image
    print(f"DEBUG: PIL imported successfully", file=sys.stderr)
except Exception as e:
    print(json.dumps({"error": f"Failed to import PIL: {str(e)}"}))
    sys.exit(2)

# -------------------------------
# CLASS NAMES (Must match training order)
# -------------------------------
CLASS_NAMES = [
    'Apple___Apple_scab',
    'Apple___Black_rot',
    'Apple___Cedar_apple_rust',
    'Apple___healthy',
    'Blueberry___healthy',
    'Cherry_(including_sour)___Powdery_mildew',
    'Cherry_(including_sour)___healthy',
    'Corn_(maize)___Cercospora_leaf_spot Gray_leaf_spot',
    'Corn_(maize)___Common_rust_',
    'Corn_(maize)___Northern_Leaf_Blight',
    'Corn_(maize)___healthy',
    'Grape___Black_rot',
    'Grape___Esca_(Black_Measles)',
    'Grape___Leaf_blight_(Isariopsis_Leaf_Spot)',
    'Grape___healthy',
    'Orange___Haunglongbing_(Citrus_greening)',
    'Peach___Bacterial_spot',
    'Peach___healthy',
    'Pepper,_bell___Bacterial_spot',
    'Pepper,_bell___healthy',
    'Potato___Early_blight',
    'Potato___Late_blight',
    'Potato___healthy',
    'Raspberry___healthy',
    'Soybean___healthy',
    'Squash___Powdery_mildew',
    'Strawberry___Leaf_scorch',
    'Strawberry___healthy',
    'Tomato___Bacterial_spot',
    'Tomato___Early_blight',
    'Tomato___Late_blight',
    'Tomato___Leaf_Mold',
    'Tomato___Septoria_leaf_spot',
    'Tomato___Spider_mites Two-spotted_spider_mite',
    'Tomato___Target_Spot',
    'Tomato___Tomato_Yellow_Leaf_Curl_Virus',
    'Tomato___Tomato_mosaic_virus',
    'Tomato___healthy'
]

# -------------------------------
# MODEL PATH (Use only .h5)
# -------------------------------
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
MODEL_PATH = os.path.join(SCRIPT_DIR, 'trained_model.h5')

print(f"DEBUG: Script directory: {SCRIPT_DIR}", file=sys.stderr)
print(f"DEBUG: Looking for model at: {MODEL_PATH}", file=sys.stderr)
print(f"DEBUG: Model exists: {os.path.exists(MODEL_PATH)}", file=sys.stderr)

_model = None

def load_model():
    """Load trained model once."""
    global _model
    if _model is None:
        if not os.path.exists(MODEL_PATH):
            error_msg = f"Model file not found at: {MODEL_PATH}"
            print(json.dumps({"error": error_msg}))
            sys.exit(2)
        try:
            print(f"DEBUG: Loading model from {MODEL_PATH}", file=sys.stderr)
            _model = keras.models.load_model(MODEL_PATH)
            print(f"DEBUG: Model loaded successfully", file=sys.stderr)
        except Exception as e:
            error_msg = f"Failed to load model: {str(e)}"
            print(json.dumps({"error": error_msg}))
            sys.exit(2)
    return _model


def preprocess_image(image_path, target_size=(128, 128)):
    """Preprocess image: resize, normalize, expand dims."""
    try:
        print(f"DEBUG: Opening image: {image_path}", file=sys.stderr)
        img = Image.open(image_path).convert('RGB')
        img = img.resize(target_size)
        arr = np.asarray(img, dtype=np.float32) / 255.0
        arr = np.expand_dims(arr, axis=0)
        print(f"DEBUG: Image preprocessed successfully", file=sys.stderr)
        return arr
    except Exception as e:
        error_msg = f"Failed to preprocess image: {str(e)}"
        print(json.dumps({"error": error_msg}))
        sys.exit(3)


def main():
    parser = argparse.ArgumentParser(description="Plant disease predictor")
    parser.add_argument("--image", type=str, required=True, help="Path to image file")
    args = parser.parse_args()

    image_path = args.image
    print(f"DEBUG: Image path argument: {image_path}", file=sys.stderr)

    if not os.path.exists(image_path):
        error_msg = f"Image not found: {image_path}"
        print(json.dumps({"error": error_msg}))
        sys.exit(1)

    try:
        model = load_model()
        inp = preprocess_image(image_path, target_size=(128, 128))  

        print(f"DEBUG: Making prediction...", file=sys.stderr)
        preds = model.predict(inp, verbose=0)

        if preds.ndim == 2:
            probs = preds[0]
        else:
            probs = np.array(preds).ravel()

        idx = int(np.argmax(probs))
        confidence = float(round(float(probs[idx]), 4))
        disease = CLASS_NAMES[idx] if idx < len(CLASS_NAMES) else f"class_{idx}"

        result = {
            "disease": disease,
            "confidence": confidence
        }
        print(f"DEBUG: Prediction complete: {disease} ({confidence})", file=sys.stderr)
        print(json.dumps(result))
        sys.exit(0)

    except Exception as e:
        error_msg = f"Prediction error: {str(e)}"
        print(json.dumps({"error": error_msg}))
        print(f"DEBUG: Exception details: {repr(e)}", file=sys.stderr)
        sys.exit(3)


if __name__ == "__main__":
    main()
